### PR TITLE
Challenge refactor and new metrics

### PR DESCRIPTION
**Pull request by joezuntz**

### CODE REVIEW

File: classifiers/trivial.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+"""
+This is an example tomographic bin generator using a random forest.
+
+Every classifier module needs to implement define _provides variable
```
Comment (joezuntz) [inline classifiers/trivial.py:4]: Same comment as above.

File: classifiers/trivial.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+that list classifiers. 
+
+See Classifier Documentation below
+"""
+import numpy as np
+
+class Random:
+    """ Completely random classifier. """
+    # Which bands are valid to call this with
+    _valid_bads = ["riz","griz"]
```
Comment (joezuntz) [inline classifiers/trivial.py:14]: same typos

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+scikit-learn
+sacc
+git+https://github.com/LSSTDESC/firecrown/
+sacc
+h5py
+matplotlib
+git+https://github.com/cmbant/camb
+cosmosis-standalone
+pyccl
+camb
```
Comment (joezuntz) [inline requirements.txt:10]: add click

File: tomo_challenge/__init__.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+from .metrics import compute_scores
+from .data import load_magnitudes_and_colors, load_redshift
```
Comment (joezuntz) [inline tomo_challenge/__init__.py:2]: we might want to rewrite to also load the size and snr metacal quantities too.

File: tomo_challenge/metrics.py [hunk 5]
```diff
@@ -22,26 +35,88 @@ def compute_snr_score(tomo_bin, z):
 
     Returns
     -------
-    score: float
-        Metric for this configuration
+    scores: dict
+         A dictionary of scores. The following dict keys are present
+
+        "SNRww", "SNRgg", "SNR3x2": float
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:41]: put underscores in these keys.

File: tomo_challenge/metrics.py [hunk 6]
```diff
@@ -22,26 +35,88 @@ def compute_snr_score(tomo_bin, z):
...
+        P = np.linalg.inv(C)
+        scores['SNR'+what] = (mu.T @ P @ mu)**0.5 
+
+        sacc_data = make_sacc(tomo_bin, z, what, mu, C)
+        scores['FOM'+what] = figure_of_merit(sacc_data, what, galaxy_galaxy_tracer_bias)
+
+    return scores
+
+
+def compute_fom_metric(tomo_bin, z):
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:63]: delete

File: classifiers/trivial.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    def train (self,training_data, training_z):
+        pass
+
+    def apply (self,data):
+        nbins = self.opt["bins"]
+        tomo_bin = np.random.randint(0,nbins,len(data))
+        return tomo_bin
+
+
+class ILover:
```
Comment (joezuntz) [inline classifiers/trivial.py:31]: Improve name.  The class doesn't love the i-band, the user does.

File: bin/challenge.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            print (f"Classifier {classifier} not found.")
+            sys.exit(1)
+        opts = json.loads("{"+options+"}")
+        run_one(classifier, classifiers[classifier], bands, opts, competition)
+
+    
+def find_modules():
+    import glob, importlib.util
+    classifiers = {}
+    for file in glob.glob(os.path.join(root_dir,"classifiers","*.py")):
```
Comment (joezuntz) [inline bin/challenge.py:33]: comment to describe what's happening here.

File: bin/challenge.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import json
+
+## get root dir, one dir below this script
+root_dir=os.path.join(os.path.split(sys.argv[0])[0],"..")
+sys.path.append(root_dir)
+import tomo_challenge as tc
+
+@click.command()
+@click.argument('action', type=str)
+@click.option('-c', '--classifier', type=str, default='RandomForest', show_default=True, help = "comma separated classifier(s)")
```
Comment (joezuntz) [inline bin/challenge.py:14]: only one is supported

File: classifiers/random_forest.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+        # for speed, cut down to 5% of original size
+        cut = np.random.uniform(0, 1, training_z.size) < 0.05
+        training_bin = training_bin[cut]
+        training_data = training_data[cut]
+
+        # Can be replaced with any classifier
+        classifier = RandomForestClassifier()
+
+        t0 = time.perf_counter()
```
Comment (joezuntz) [inline classifiers/random_forest.py:56]: delete timing - not used here.

File: classifiers/trivial.py (hunk 7 above)
Comment (joezuntz) [reply classifiers/trivial.py:31]: Also move to own file

File: classifiers/trivial.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        self.indx = bands.find("i")
+        self.opt = options
+
+    def train (self,training_data, training_z):
+        pass
+
+    def apply (self,data):
+        nbins = self.opt["bins"]
+        iband = data[:,self.indx]
+        tomo_bin = np.digitize(iband, np.linspace(iband.min(),iband.max(), nbins))
```
Comment (joezuntz) [inline classifiers/trivial.py:50]: isn't there an edge case here with the highest or lowest object?

File: tomo_challenge/config.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    step_size: 0.02
+  parameters:
+    Omega_c: [0.25, 0.27, 0.32]
+    Omega_b: [0.04, 0.045, 0.05]
+    h: [0.5, 0.67, 0.9]
+    sigma8: [0.74, 0.8404844953840714, 0.94]
+    n_s: [0.9,  0.96,  1.02]
+    # I'd like to do the proper DETF FoM,
+    # but right now our camb config rejects this
+    # since it crosses w=-1.  So for now use am
```
Comment (joezuntz) [inline tomo_challenge/config.yml:30]: typo

File: tomo_challenge/metrics.py [hunk 13]
```diff
@@ -69,51 +148,63 @@ def compute_mean_covariance(tomo_bin, z):
...
-    z_mid = (z_edge[1:] + z_edge[:-1]) / 2
-
-
-    # Generate CCL tracers and get total counts, for the noise
-    counts = np.zeros(nbin)
+    z_mid, n_of_z = get_n_of_z(tomo_bin, z)
+    bz = 1/ccl.growth_factor(cosmo, 1/(1+z_mid))
+    bofz=(z_mid,bz)
+    galaxy_galaxy_tracer_bias = [(bz*nz_bin).sum()/(nz_bin).sum() for nz_bin in n_of_z]
+    print ('biases=',galaxy_galaxy_tracer_bias)
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:155]: use f-string syntax (cleaner)

File: tomo_challenge/metrics.py [hunk 14]
```diff
@@ -69,51 +148,63 @@ def compute_mean_covariance(tomo_bin, z):
...
         # Noise contribution, if an auto-bin
-        if i == j:
-            C_obs[i, j] = C_sig[i, j] + sigma_e**2 / n_eff[i]
+        if ci == cj:
+            if tracertype[i]=='g':
+                C_obs[ci, cj] = C_sig[ci, cj] + 1/n_bar[i]
+            elif tracertype[i]=='w':
+                C_obs[ci, cj] = C_sig[ci, cj] + sigma_e**2 / n_eff[i]
+            else:
+                print ("Prasec!")
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:203]: ?

File: tomo_challenge/metrics.py [hunk 15]
```diff
@@ -69,51 +148,63 @@ def compute_mean_covariance(tomo_bin, z):
...
+    ## start with number counts
+    if (what=='gg' or what=='3x2'):
+        tracers += [ccl.NumberCountsTracer(cosmo, has_rsd=False, dndz=(z_mid, nz_bin), bias=bofz)
+                    for nz_bin in n_of_z]
+    if (what=='ww' or what=='3x2'):
+        tracers += [ccl.WeakLensingTracer(cosmo, dndz=(z_mid, nz_bin))
+                    for nz_bin in n_of_z] 
 
+    ntracers = len(tracers)
+    print (tracertype)
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:170]: remove printout

File: tomo_challenge/metrics.py [hunk 16]
```diff
@@ -69,51 +148,63 @@ def compute_mean_covariance(tomo_bin, z):
...
     # and the version with noise (for the covmat)
     C_sig = {}
     C_obs = {}
-    for i, j in blocks:
-        Ti = tracers[i]
-        Tj = tracers[j]
-        C_sig[i, j] = ccl.angular_cl(cosmo, Ti, Tj, ell)
-
+    for ci, cj in blocks:
+        i = ci % nbin
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:191]: explain why % here in comments

File: tomo_challenge/metrics.py [hunk 17]
```diff
@@ -22,26 +35,88 @@ def compute_snr_score(tomo_bin, z):
...
+def ell_binning():
+    # we put this here to make sure it's used consistently
+    # plausible limits I guess
+    ell_max = 2000
+    n_ell = 100
+    # choose ell bins from 10 .. 2000 log spaced
+    ell = np.logspace(2, np.log10(ell_max), n_ell)
+    return ell
+
+def get_tracertype (nbin,what):
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:99]: add underscore in name

Comment (joezuntz) [reply tomo_challenge/metrics.py:99]: get_tracer_type

File: tomo_challenge/metrics.py [hunk 18]
```diff
@@ -154,3 +245,141 @@ def plot_distributions(z, tomo_bin, filename, nominal_edges=None):
...
+
+    # Add all the tracers
+    for ci in range(ntot):
+        i = ci %nbin
+        if tracertype[ci]=='g':
+            S.add_tracer("NZ", f'lens_{i}', z_mid, n_of_z[i])
+        elif tracertype[ci]=='w':
+            S.add_tracer("NZ", f'source_{i}', z_mid, n_of_z[i])
+        else:
+            raise NotImplemented
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:278]: This should be `NotImplementedError` 
(`NotImplemented` is for overriding mathematical operators in classes)

File: tomo_challenge/metrics.py [hunk 19]
```diff
@@ -154,3 +245,141 @@ def plot_distributions(z, tomo_bin, filename, nominal_edges=None):
...
+
+    # Now run firecrown in a temporary directory
+    # and load the resulting fisher matrix
+    conf, data = firecrown.parse(config)
+    with tempfile.TemporaryDirectory() as tmp:
+        firecrown.run_cosmosis(conf, data, pathlib.Path(tmp))
+        fisher_matrix = np.loadtxt(f'{tmp}/chain.txt')
+
+    # Pull out the correct indices.  We would like to use
+    # w0 - wa but can't yet because
```
Comment (joezuntz) [inline tomo_challenge/metrics.py:369]: complete comment

**Comment by slosar** [comment]
Move to yaml config, it makes more sense, I think. More soon.

**Pull request merged**